Gawron/ZPI-42-ZPI-64-BE-39-BE-58-Rezerwacja-i-zarządzanie-opieką

Gawron97 opened the pull request:
Dodanie logiki dodania rezerwacji (jako opieka), edytowania opieki, zmieniania statusu.

Gawron97 committed:
add endpoints for making reservation for care and for updating care by caretaker

Gawron97 committed:
add endpoints for changing statuses

Gawron97 committed:
add integration tests for CareController and CareService

Gawron97 committed:
add integration tests for AnimalService and add Unit test for CareMapper

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
Chyba brakuje `@PreAuthorize("isAuthenticated()")`

lukcyn replied:
Też zastanawia mnie czy tutaj nie pododawać jakiegoś headera jeśli endpoint jest tylko dla klienta/opiekuna

Na swoim branchu mam
`@RequestHeader(value = "${header-name.role}") Role acceptRole`
Tylko z racji tego że websockety są problematyczne to jeszcze mi troche zajmie zanim to wrzuce

Gawron97 replied:
ja mam informacje w opiece, kto jest kto, wiec dla bezpieczenstwa i tak będe to sprawdzal, wiec chyba nie ma sensu u mnie

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -0,0 +1,49 @@
+    private final CareService careService;
+
+    @PostMapping("/reservation")
+    public CareDTO makeReservation(@RequestBody CreateCareDTO createCare, Principal principal) {
```
Wydaje mi się, że bez `@Valid` nie zvaliduje tego obiektu

Gawron97 replied:
aaaa, rzeczywiscie

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PatchMapping("/{careId}/update")
+    public CareDTO updateCare(@PathVariable Long careId, @RequestBody UpdateCareDTO updateCare, Principal principal) {
```
Jak wyżej

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/care/CreateCareDTO.java:
```diff
@@ -0,0 +1,22 @@
+        @NotNull LocalDate careStart,
+        @NotNull LocalDate careEnd,
+        String description,
+        @NotNull BigDecimal dailyPrice,
```
Można by dodać jeszcze walidację żeby to było większe od zera. Np. znalazłem coś takiego

```java
    @DecimalMin(value = "0.0", inclusive = false)
    @Digits(integer=10, fraction=2)
    @NotNull 
    BigDecimal dailyPrice
```

Gawron97 replied:
a w sumie, teraz daje mozliwosc ze klient doplaci jak student na praktykach :)

lukcyn replied:
Czyli zostawiamy ten feature? :))

Gawron97 replied:
oczywiscie 💸

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/care/UpdateCareDTO.java:
```diff
@@ -0,0 +1,15 @@
+public record UpdateCareDTO(
+        @NotNull LocalDate careStart,
+        @NotNull LocalDate careEnd,
+        @NotNull BigDecimal dailyPrice
```
Można by dodać jeszcze walidację żeby to było większe od zera. Np. znalazłem coś takiego

```java
    @DecimalMin(value = "0.0", inclusive = false)
    @Digits(integer=10, fraction=2)
    @NotNull 
    BigDecimal dailyPrice
```

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/care/CareDTO.java:
```diff
@@ -0,0 +1,25 @@
+        String animalType,
+        Map<String, List<String>> selectedOptions,
+        String caretakerEmail,
+        String clientEmail
```
W sumie dlaczego to tu jest skoro to samo jest w Principal?

Gawron97 replied:
to jest DTO zwracane dla frontu z informacjami o opiece

lukcyn replied:
A fakt. Chyba mi chodziło bardziej żeby zaznaczyć CreateCareDTO. Ale to już napisałeś wcześniej dlaczego tak jest więc czaję

lukcyn commented on src/main/java/com/example/petbuddybackend/service/care/CareService.java:
```diff
@@ -0,0 +1,203 @@
+
+    public CareDTO makeReservation(CreateCareDTO createCare, String clientEmail) {
+
+        assertLoggedInUserIsClient(createCare.clientEmail(), clientEmail,
```
To teraz pytanie dlaczego ten check jest tu potrzebny skoro wiadomo, że to principal robi request?

Gawron97 replied:
tylko, ze princinpalem moze byc ktos inny, np. uzytkownik z profilu opiekuna zbuduje request reczenie( bo front i tam uniemowliwi )

lukcyn replied:
To wtedy mógłby być ten header z rolą przydatny

Gawron97 replied:
wydaje mi sie ze jak jest mozliwosc sprawdzenia 1 ifem to dla frontu wygodniej

lukcyn commented on src/main/java/com/example/petbuddybackend/service/care/CareService.java:
```diff
@@ -0,0 +1,203 @@
+        assertLoggedInUserIsClient(createCare.clientEmail(), clientEmail,
+                "Client can only make reservation for themselves");
+        assertClientIsNotCaretaker(createCare.clientEmail(), createCare.caretakerEmail());
+        assertEndCareDateIsAfterStartCareDate(createCare.careStart(), createCare.careEnd());
```
Można jeszcze dać assert że obie daty są z przyszłości

lukcyn commented on src/main/java/com/example/petbuddybackend/service/care/CareService.java:
```diff
@@ -0,0 +1,203 @@
+                .orElseThrow(() -> new NotFoundException("Care not found"));
+    }
+
+    public CareDTO acceptCareByCaretaker(Long careId, String userEmail) {
+
+        Care care = getCare(careId);
+        assertLoggedInUserIsCaretaker(care.getCaretaker().getEmail(), userEmail,
+                "Only caretaker can change caretaker status");
+        assertCareIsNotTerminated(care);
+        assertCaretakerStatusIsPending(care, "Care already accepted");
+        assertClientStatusIsAccepted(care);
+        care.setCaretakerStatus(CareStatus.AWAITING_PAYMENT);
+        care.setClientStatus(CareStatus.AWAITING_PAYMENT);
+        return careMapper.mapToCareDTO(careRepository.save(care));
+    }
```
Zastanawia mnie, czy nie prościej byłoby zrobić z tego maszynę stanu tj. zdefiniować jaki stan może nastąpić po jakim a w przeciwnym wypadku rzucić wyjątek

pseudokodem byłoby np.

```
Stan(<Lista stanów które mogą nastąpić po tym stanie>)
PENDING(ACCEPTED, CANCELLED, OUTDATED)
AWAITING_PAYMENT(PAID, OUTDATED, CANCELLED)
```

Na pewno to byłoby fajne do PrePresist żeby głupot nie powpisywać

Gawron97 replied:
zastanawialem sie nad tym, ale nigdy nie robilem czegos takiego, zerkne, chociaz wydaje mi sie ze to wszystko i tak jest w kazdym scenariuszu do sprawdzenia 1 ifem

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.clientStatus").value(CareStatus.ACCEPTED.name()))
+                .andExpect(jsonPath("$.caretakerStatus").value(CareStatus.PENDING.name()))
+//                .andExpect(jsonPath("$.careStart").value(LocalDate.now().plusDays(2)))
+//                .andExpect(jsonPath("$.careEnd").value(LocalDate.now().plusDays(7)))
```
Zakomentowany kod bym wyrzucił

Gawron97 replied:
aa dobrze ze przypominasz, ja go mialem naprawic xD bo mi zwracalo date w formie tablicy xd

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+                                "email",
+                                "clientEmail")))
+                .andExpect(status().isBadRequest());
+    }
```
W sumie to serwis rzuca więc imo mogłoby to być testowane w testach serwisu

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+
+    @Test
+    @WithMockUser(username = "email")
+    void makeReservation_WhenClientEmailEqualsCaretakerEmail_ThenThrowIllegalActionException() throws Exception {
+        // When and Then
+        mockMvc.perform(post("/api/care/reservation")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(String.format(CREATE_CARE_BODY,
+                                LocalDate.now().plusDays(2),
+                                LocalDate.now().plusDays(7),
+                                "Test care description",
+                                new BigDecimal("50.00"),
+                                "DOG",
+                                "",
+                                "email",
+                                "email")))
+                .andExpect(status().isBadRequest());
+    }
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+                                "email",
+                                "clientEmail")))
+                .andExpect(status().isBadRequest());
+    }
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+
+    @Test
+    @WithMockUser(username = "clientEmail")
+    void updateCare_WhenClientTriesToEdit_ShouldThrowIllegalActionException() throws Exception {
+        // Given
+        PersistenceUtils.addCare(careRepository, caretaker, client, animalRepository.findById("DOG").get());
+        // When and Then
+        Long careId = careRepository.findAll().get(0).getId();
+        mockMvc.perform(patch("/api/care/" + careId + "/update")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(String.format(UPDATE_CARE_BODY,
+                                LocalDate.now().plusDays(3),
+                                LocalDate.now().plusDays(8),
+                                new BigDecimal("60.00")))
+                )
+                .andExpect(status().isBadRequest());
+    }
```
Może być przeniesione do testu serwisu

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+
+    @Test
+    @WithMockUser(username = "email")
+    void updateCare_WhenCareIsCancelled_ShouldThrowIllegalActionException() throws Exception {
+        // Given
+        Care care = PersistenceUtils.addCare(careRepository, caretaker, client, animalRepository.findById("DOG").get());
+        care.setClientStatus(CareStatus.CANCELLED);
+        careRepository.save(care);
+        // When and Then
+        Long careId = careRepository.findAll().get(0).getId();
+        mockMvc.perform(patch("/api/care/" + careId + "/update")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(String.format(UPDATE_CARE_BODY,
+                                LocalDate.now().plusDays(3),
+                                LocalDate.now().plusDays(8),
+                                new BigDecimal("60.00")))
+                )
+                .andExpect(status().isBadRequest());
+    }
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+                .andExpect(jsonPath("$.clientStatus").value(CareStatus.PENDING.name()))
+                .andExpect(jsonPath("$.caretakerStatus").value(CareStatus.PENDING.name()))
+//                .andExpect(jsonPath("$.careStart").value(LocalDate.now().plusDays(2)))
+//                .andExpect(jsonPath("$.careEnd").value(LocalDate.now().plusDays(7)))
```
Komentarze do usunięcia

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+
+    @Test
+    @WithMockUser(username = "email")
+    void updateCare_WhenCareIsAcceptedByCaretaker_ShouldThrowIllegalActionException() throws Exception {
+        // Given
+        Care care = PersistenceUtils.addCare(careRepository, caretaker, client, animalRepository.findById("DOG").get());
+        care.setCaretakerStatus(CareStatus.ACCEPTED);
+        careRepository.save(care);
+        // When and Then
+        Long careId = careRepository.findAll().get(0).getId();
+        mockMvc.perform(patch("/api/care/" + careId + "/update")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(String.format(UPDATE_CARE_BODY,
+                                LocalDate.now().plusDays(3),
+                                LocalDate.now().plusDays(8),
+                                new BigDecimal("60.00")))
+                )
+                .andExpect(status().isBadRequest());
+    }
```
Może być przeniesione do testów serwisu

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+                                new BigDecimal("60.00")))
+                )
+                .andExpect(status().isBadRequest());
+    }
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+        Long careId = careRepository.findAll().get(0).getId();
+        mockMvc.perform(post("/api/care/" + careId + "/caretaker-accept"))
+                .andExpect(status().isBadRequest());
+    }
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -0,0 +1,487 @@
+    }
+
+
+}
```
Jak wyżej

lukcyn commented on src/test/java/com/example/petbuddybackend/service/animal/AnimalServiceIntegrationTest.java:
```diff
@@ -0,0 +1,68 @@
+
+@SpringBootTest
+@ContextConfiguration(classes = TestDataConfiguration.class)
+public class AnimalServiceIntegrationTest {
```
Nie powiedziałbym że to jest test integracyjny

Gawron97 replied:
jest, bo jeszcze korzysta z bazy :)

lukcyn commented on src/test/java/com/example/petbuddybackend/service/care/CareServiceIntegrationTest.java:
```diff
@@ -0,0 +1,514 @@
+
+@SpringBootTest
+@ContextConfiguration(classes = TestDataConfiguration.class)
+public class CareServiceIntegrationTest {
```
Nie powiedziałbym że to jest test integracyjny

lukcyn reviewed (commented):
Pare komentarzy i pytań

W sumie byłbym za refaktorem tych testów bo imo w testach kontrolera nie trzeba testować serwisu przy okazji ale sam kontroler

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/care/CareDTO.java:
```diff
@@ -0,0 +1,25 @@
+
+public record CareDTO(
+        Long id,
+        ZonedDateTime submittedAt,
+        CareStatus caretakerStatus,
+        CareStatus clientStatus,
+        LocalDate careStart,
+        LocalDate careEnd,
```
Można jeszcze do dat dodać `@JsonFormat(pattern =...)
`

Gawron97 replied:
Tak tak

Gawron97 committed:
Formatted date for user zone
Validating start and end date of care
fix tests

Gawron97 committed:
add documentation and secure endpoints

Gawron97 committed:
exclude config.annotations from test coverage
change exception message

Gawron97 committed:
checking in DateRangeValidator for proper using on date fields

Gawron97 committed:
schedule for terminate cares at 00:01

Gawron97 committed:
add CareScheduledTest

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/annotations/validation/DateRangeValidator.java:
```diff
@@ -27,14 +30,22 @@ public boolean isValid(Object value, ConstraintValidatorContext context) {
             }
 
-            boolean isValid = startDate.compareTo(endDate) < 0;
+            if(!(startDate instanceof Temporal) || !(endDate instanceof Temporal) ||
+                    !(startDate instanceof Comparable) || !(endDate instanceof Comparable)) {
+                throw new IllegalActionException("DateRangeValidator can only be used on Date fields implementing Comparable");
+            }
+
+            Comparable<Object> startDateComparable = (Comparable<Object>) startDate;
+            Comparable<Object> endDateComparable = (Comparable<Object>) endDate;
+
+            boolean isValid = startDateComparable.compareTo(endDateComparable) < 0;
```
W sumie isValid to może być funkcja osobna

```java
    private static boolean isValid(Object startDate, Object endDate) {
        if(!(startDate instanceof Temporal) || !(endDate instanceof Temporal) ||
                !(startDate instanceof Comparable) || !(endDate instanceof Comparable)) {
            throw new IllegalActionException("DateRangeValidator can only be used on Date fields implementing Comparable");
        }

        Comparable<Object> startDateComparable = (Comparable<Object>) startDate;
        Comparable<Object> endDateComparable = (Comparable<Object>) endDate;

        return startDateComparable.compareTo(endDateComparable) < 0;
    }
```

Gawron97 replied:
w sumie

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/scheduled/CareScheduled.java:
```diff
@@ -0,0 +1,38 @@
+package com.example.petbuddybackend.utils.scheduled;
```
Nie bardzo pasuje do utilsów, dałbym wyżej

Gawron97 replied:
moze tak byc

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/scheduled/CareScheduled.java:
```diff
@@ -0,0 +1,38 @@
+        List<Care> cares = careRepository.findAllByCaretakerStatusNotInOrClientStatusNotIn(
+                List.of(CareStatus.CANCELLED, CareStatus.OUTDATED, CareStatus.PAID),
+                List.of(CareStatus.CANCELLED, CareStatus.OUTDATED, CareStatus.PAID));
+        log.info("Terminating cares from potential: ", cares.size());
```
Warning: More arguments provided (1) than placeholders specified (0) 

Chyba trzeba `{}` dodać

Gawron97 committed:
merge with main
change packaging
add TimeZoneParameter to Care endpoints
split method in DateRangeValidator

Gawron97 committed:
add create-drop and orphanRemoval
change package naming

Gawron97 committed:
change code coverage packages

Gawron97 committed:
change package name
